### PR TITLE
Case: OCP-32697 Forward logs to different kafka topics

### PR DESCRIPTION
**Pull request by anpingli**
https://mastern-jenkins-csb-openshift-qe.cloud.paas.psi.redhat.com/job/Runner-v3/209970/console

### CODE REVIEW

File: features/logging/cluster_log_forwarder.feature [hunk 1]
```diff
@@ -295,3 +295,50 @@ Feature: cluster log forwarder features
...
+    # The following step will create 4 topics(topic-logging-all,topic-logging-infra,topic-logging-app,topic-logging-audit)
+    Given I deployed kafka in the "<%= cb.kafka_project.name %>" project via amqstream operator
+    And I run the :extract client command with:
+      | resource | secret/my-cluster-cluster-ca-cert  |
+    Then the step should succeed
+    Given I use the "openshift-logging" project
+    Given admin ensures "instance" cluster_log_forwarder is deleted from the "openshift-logging" project after scenario
+    Given admin ensures "kafka-fluent" secret is deleted from the "openshift-logging" project after scenario
+    When I run the :create_secret client command with:
+      | secret_type | generic    |
```
Comment (pruan-rht) [inline features/logging/cluster_log_forwarder.feature:317]: @anpingli please align all tables.

File: features/step_definitions/logging.rb [hunk 2]
```diff
@@ -705,3 +705,135 @@
...
+# This step will deploy one kafka cluster and create 4 topics(topic-logging-all,topic-logging-infra,topic-logging-app,topic-logging-audit) in this kafka
+Given /^I deployed kafka in the #{QUOTED} project via amqstream operator$/ do | project_name|
+  ensure_admin_tagged
+  step %Q/I switch to cluster admin pseudo user/
+  step %Q/I use the "#{project_name}" project/
+  step %Q/I process and create:/, table(%{
+    | f | #{BushSlicer::HOME}/testdata/logging/clusterlogforwarder/kafka/amq/02_og_amqstreams_template.yaml |
+    | p | AMQ_NAMESPACE=#{project_name} |
+  })
+  raise "Error create operatorgroup" unless @result[:success]
```
Comment (QiaolingTang) [inline features/step_definitions/logging.rb:718]: I don't know if the step will work or not as I don't find where the `@result` come from. I suggest to use `step %Q/the step should succeed/`

File: features/step_definitions/logging.rb [hunk 3]
```diff
@@ -705,3 +705,135 @@
...
+
+  step %Q/I process and create:/, table(%{
+    | f | #{BushSlicer::HOME}/testdata/logging/clusterlogforwarder/kafka/amq/03_sub_amqstreams_template.yaml |
+    | p | AMQ_NAMESPACE=#{project_name} |
+  })
+  raise "Error subscript amqstreams" unless @result[:success]
```
Comment (QiaolingTang) [inline features/step_definitions/logging.rb:724]: Same comment as line 718

File: features/step_definitions/logging.rb [hunk 4]
```diff
@@ -705,3 +705,135 @@
...
+
+  step %Q/a pod becomes ready with labels:/, table(%{
+    | name=amq-streams-cluster-operator |
+  })
+
+  step %Q/I process and create:/, table(%{
+    | f | #{BushSlicer::HOME}/testdata/logging/clusterlogforwarder/kafka/amq/04_kafka_my-cluster_amqstreams_template.yaml |
+    | p | AMQ_NAMESPACE=#{project_name} |
+  })
+  raise "Error create kafka" unless @result[:success]
```
Comment (QiaolingTang) [inline features/step_definitions/logging.rb:734]: Same comment as line 718

File: features/step_definitions/logging.rb [hunk 5]
```diff
@@ -705,3 +705,135 @@
...
+
+  @result = admin.cli_exec(:create, f: "#{BushSlicer::HOME}/testdata/logging/clusterlogforwarder/kafka/amq/05_kafkatopics_amqstreams.yaml")
+  raise "Error create kafka topics" unless @result[:success]
+
+  step %Q/3 pods becomes ready with labels:/, table(%{
+    | strimzi.io/name=my-cluster-kafka |
+  })
+  raise "Error kafka cluster not ready" unless @result[:success]
```
Comment (QiaolingTang) [inline features/step_definitions/logging.rb:742]: Same comment as line 718

File: features/step_definitions/logging.rb [hunk 6]
```diff
@@ -705,3 +705,135 @@
...
+#    | image   | #{kafka_image} |
+#    | restart | Never |
+#    | command | true  |
+#    | cmd     | bin/kafka-console-consumer.sh  |
+#    | cmd     | --bootstrap-server my-cluster-kafka-bootstrap:9092  |
+#    | cmd     | --topic #{topic_name}   |
+#    | cmd     | --from-beginning  |
+#    | cmd     | --max-messages #{record_num} |
+#    | cmd     | --timeout-ms=300000  |
+#})
```
Comment (QiaolingTang) [inline features/step_definitions/logging.rb:760]: If you don't need the step, please remove it.

File: features/step_definitions/logging.rb [hunk 7]
```diff
@@ -705,3 +705,135 @@
     end
   end
 end
+
+# This step will deploy one kafka cluster and create 4 topics(topic-logging-all,topic-logging-infra,topic-logging-app,topic-logging-audit) in this kafka
+Given /^I deployed kafka in the #{QUOTED} project via amqstream operator$/ do | project_name|
```
Comment (QiaolingTang) [inline features/step_definitions/logging.rb:710]: `deployed` --> `deploy`

File: features/step_definitions/logging.rb [hunk 8]
```diff
@@ -705,3 +705,135 @@
...
+#    | cmd     | --max-messages #{record_num} |
+#    | cmd     | --timeout-ms=300000  |
+#})
+Given /^I get(?: (\d+))? records from the #{QUOTED} kafka topic$/ do | record_num, topic_name|
+  record_num = record_num ? record_num.to_str : "10"
+  job_name=topic_name
+  kafka_image=stateful_set('my-cluster-kafka').containers_spec(user: user)[0].image
+  retries = 3
+  while true
+    step %Q/I process and create:/,table(%{
```
Comment (QiaolingTang) [inline features/step_definitions/logging.rb:767]: Which project is the step working on?

File: features/step_definitions/logging.rb [hunk 9]
```diff
@@ -705,3 +705,135 @@
...
+    })
+    step %Q/a pod becomes ready with labels:/, table(%{
+      | job-name=#{job_name}|
+    })
+    sleep 30
+    step %Q/I run the :logs client command with:/, table(%{
+      | resource_name | #{ pod.name} |
+    })
+
+    admin.cli_exec(:delete, object_type: 'job', object_name_or_id: job_name, n: cb.kafka_project.name)
```
Comment (QiaolingTang) [inline features/step_definitions/logging.rb:782]: If any one of the above steps failed, this step couldn't be executed, then the whole step will fail.

File: features/step_definitions/logging.rb [hunk 10]
```diff
@@ -705,3 +705,135 @@
...
+    if @result[:response].match("pipeline_metadata")
+      @result[:success] = true
```
Comment (QiaolingTang) [inline features/step_definitions/logging.rb:784]: What's the usage of `@result[:success]` ?

File: features/step_definitions/logging.rb [hunk 11]
```diff
@@ -705,3 +705,135 @@
...
+    end
+    sleep 30
+    retries -= 1
+  end
+end
+
+# Register an topic consumer which listen the topic continuously. we can verify the logging records by the next step 'I get N logs from the X kafka consumer job'
+Given /^I listen on the #{QUOTED} kakafa topic using the #{QUOTED} job$/ do | topic_name, job_name |
+  kafka_image=statefulset('my-cluster-kafka').containers_spec(user: user)[0].image
+  step %Q/I process and create:/,table(%{
```
Comment (QiaolingTang) [inline features/step_definitions/logging.rb:798]: Please add clean-up steps.

File: features/step_definitions/logging.rb [hunk 12]
```diff
@@ -705,3 +705,135 @@
...
+    | f | "#{BushSlicer::HOME}/testdata/logging/clusterlogforwarder/kafka/amq/21_job_topic_consumer_from_latest_template.yaml" |
+    | p | KAFKA_IMAGE=#{kafka_image} |
+    | p | KAFKA_TOPIC=#{topic_name} |
+    | p | MAX_MESSAGES=#{record_num} |
+    | p | JOB_NAME=#{job_name} |
+  })
+end
+
+# Get logs from the kafka comsumer job which created by upper job
+Given /^I get(?: (\d+))? logs from the #{QUOTED} kafka consumer job$/ do | record_num, job_name |
```
Comment (QiaolingTang) [inline features/step_definitions/logging.rb:808]: I have same questions as step `I get(?: (\d+))? records from the #{QUOTED} kafka topic`

File: testdata/logging/clusterlogforwarder/kafka/amq/01_namespace_amqstreams_template.yaml [hunk 13]
```diff
@@ -0,0 +1,21 @@
+kind: Template
```
Comment (QiaolingTang) [inline testdata/logging/clusterlogforwarder/kafka/amq/01_namespace_amqstreams_template.yaml:1]: Seems you don't use this file, please remove it.

File: testdata/logging/clusterlogforwarder/kafka/amq/06_kafkabridges_amqstreams.yaml [hunk 14]
```diff
@@ -0,0 +1,95 @@
+---
```
Comment (QiaolingTang) [inline testdata/logging/clusterlogforwarder/kafka/amq/06_kafkabridges_amqstreams.yaml:1]: The file isn't used, please remove it.

File: features/step_definitions/logging.rb [hunk 15]
```diff
@@ -705,3 +705,135 @@
...
+    step %Q/I run the :logs client command with:/, table(%{
+      | resource_name | #{ pod.name} |
+    })
+
+    admin.cli_exec(:delete, object_type: 'job', object_name_or_id: job_name, n: cb.kafka_project.name)
+    if @result[:response].match("pipeline_metadata")
+      @result[:success] = true
+      break
+    elsif retries == 0
+      @result[:success] = false
```
Comment (QiaolingTang) [inline features/step_definitions/logging.rb:787]: What's the usage of `@result[:success]` ? Where does it come from?
I think you can raise some error here.
`raise "error message"`

File: features/logging/cluster_log_forwarder.feature (hunk 1 above)
Comment (anpingli) [reply features/logging/cluster_log_forwarder.feature]: Done

**Comment by anpingli** [comment]
@QiaolingTang  https://mastern-jenkins-csb-openshift-qe.cloud.paas.psi.redhat.com/job/Runner-v3/211613/console

File: features/step_definitions/logging.rb [hunk 16]
```diff
@@ -705,3 +705,114 @@
...
+    | p | JOB_NAME=#{job_name} |
+  })
+  step %Q/a pod becomes ready with labels:/, table(%{
+    | job-name=#{job_name}|
+  })
+
+  # wait up to 1 minutes for the kafka message
+  success = wait_for(60, interval: 10) {
+    @result = user.cli_exec(:logs, {resource_name: "#{pod.name}"})
+    if @result[:response].match("pipeline_metadata")
```
Comment (QiaolingTang) [inline features/step_definitions/logging.rb:772]: Line 772~776 can simplify to one line `@result[:response].match("pipeline_metadata")`

File: features/step_definitions/logging.rb [hunk 17]
```diff
@@ -705,3 +705,114 @@
...
+      @result[:success] = true
+    else
+      @result[:success] = false
+    end
+  }
```
Comment (QiaolingTang) [inline features/step_definitions/logging.rb:777]: need to add something below line 777:
```
raise "xxxx" unless success
```

File: features/step_definitions/logging.rb [hunk 18]
```diff
@@ -705,3 +705,114 @@
...
+
+  # wait up to 1 minutes for the kafka message
+  success = wait_for(60, interval: 10) {
+    if record_num >0
+      @result = user.cli_exec(:logs, {resource_name: "#{pod.name}", tail: "#{ record_num }"})
+    else
+      @result = user.cli_exec(:logs, {resource_name: "#{pod.name}"})
+    end
+
+    if @result[:response].match("pipeline_metadata")
```
Comment (QiaolingTang) [inline features/step_definitions/logging.rb:812]: Same suggestions as above.

**Comment by anpingli** [comment]
@liangxia  PTAL

**Pull request merged**